tools/db/build/build_models_script.php

```diff
@@ -0,0 +1,330 @@
+        if(!isset($model->related)) continue;
+        $relatedobj = get_object_vars($model->related);
+        foreach($relatedobj as $id => $model) {
+          $deprecates = isset($related->deprecates) && $related->deprecates;
+          $result .= <<<END
+$comma
+          ({$this->sqlv($model, 'id')},
+          {$this->sqlv(null, $id)},
+          {$this->sqlb($deprecates)})
+END;
+          $comma = ',';
+        }
+      }
+      
+      if($comma == '') return ''; // no entries found
+      return $result . ";\n";
+    }
+    
+    /**
+      Safe-quotes a SQL string
+    */
+    function sqlv($o, $s) {
+      if($o !== null) {
+        if(isset($o->$s)) $s = $o->$s; 
+        else return 'null';
+      }
+      if($s === null) return 'null';
+      
+      $v = strpos($s, "\0");
+      if($v !== FALSE) {
+        $s = substr($s, 0, strpos($s, "\0"));
+      }
+      $s = iconv("UTF-8", "UTF-8//IGNORE", $s); // Strip invalid UTF-8 characters
+      //return "'" . mysql_real_escape_string($s) . "'";
+      return 
+        "'" . 
+        str_replace(["'",   "\"",   "\r",  "\n",  "\b",  "\t"], 
+                    ["\\'", "\\\"", "\\r", "\\n", "\\b", "\\t"],  
+                    str_replace("\\", "\\\\", $s) ) . 
+        "'";
+    }
+
+    /**
+      Safe-quotes a SQL date
+    */
+    function sqld($o, $s) {
+      if(isset($o->$s)) $s = $o->$s; 
+      else return 'null';
+      if($s === null) return 'null';
+      $s = substr($s, 0, 19);
+      return $this->sqlv(null, $s);
+    }
+    
+    function sqli($o, $s) {
+      if(isset($o->$s)) $s = $o->$s; 
+      else return 'null';
+      
+      if(!is_numeric($s)) die('Expecting numeric $s');
+      return $s;
+    }
+    
+    function sqlb($b) {
+      return $b ? '1' : '0';
+    }
+    
+    function parse_bcp47($bcp47, &$lang, &$region, &$script) {
+      $lang = null;
+      $region = null;
+      $script = null;
+      
+      // RegEx from https://stackoverflow.com/questions/7035825/regular-expression-for-a-language-tag-as-defined-by-bcp47, https://stackoverflow.com/a/34775980/1836776
+      $re = preg_match("/^(?<grandfathered>(?:en-GB-oed|i-(?:ami|bnn|default|enochian|hak|klingon|lux|mingo|navajo|pwn|t(?:a[oy]|su))|sgn-(?:BE-(?:FR|NL)|CH-DE))|(?:art-lojban|cel-gaulish|no-(?:bok|nyn)|zh-(?:guoyu|hakka|min(?:-nan)?|xiang)))|(?:(?<language>(?:[A-Za-z]{2,3}(?:-(?<extlang>[A-Za-z]{3}(?:-[A-Za-z]{3}){0,2}))?)|[A-Za-z]{4}|[A-Za-z]{5,8})(?:-(?<script>[A-Za-z]{4}))?(?:-(?<region>[A-Za-z]{2}|[0-9]{3}))?(?:-(?<variant>[A-Za-z0-9]{5,8}|[0-9][A-Za-z0-9]{3}))*(?:-(?<extension>[0-9A-WY-Za-wy-z](?:-[A-Za-z0-9]{2,8})+))*)(?:-(?<privateUse>x(?:-[A-Za-z0-9]{1,8})+))?$/Di", $bcp47, $matches);
+      if($re === FALSE) {
+        return false;
+      }
+      
+      if(isset($matches['language'])) $lang = $matches['language'];
+      if(isset($matches['region'])) $region = $matches['region'];
+      if(isset($matches['script'])) $script = $matches['script'];
+      return true;
+    }
```

Comment:
Could `parse_bcp47` be useful for keyboard queries too? like in a shared script?

Reply:
It could be moved to a shared script ... maybe later :)